File: create_auth_users.py
#!/usr/bin/env python3
"""
Create auth users in Supabase for sample data
This script uses Supabase Admin API to create users that match profiles.csv

Requirements:
- supabase-py: pip install supabase
- Set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables

Usage:
1. Generate profiles.csv: python generate_sample_data.py
2. Set environment variables:
   export SUPABASE_URL="your-project-url"
   export SUPABASE_SERVICE_ROLE_KEY="your-service-role-key"
3. Run this script: python create_auth_users.py
"""

import csv
import os
import sys
import json

# Try to load .env file if python-dotenv is available
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass  # python-dotenv not installed, continue without it

# Try to use supabase client, but fallback to requests if needed
try:
    from supabase import create_client, Client
    USE_SUPABASE_CLIENT = True
except ImportError:
    USE_SUPABASE_CLIENT = False
    try:
        import requests
    except ImportError:
        print("❌ Error: Need either 'supabase' or 'requests' package")
        print("   Install with: pip install supabase")
        print("   Or: pip install requests")
        sys.exit(1)

# Get Supabase credentials from environment
# Try multiple possible variable names
SUPABASE_URL = (
    os.getenv("VITE_SUPABASE_URL") or 
    os.getenv("SUPABASE_URL") or
    os.getenv("NEXT_PUBLIC_SUPABASE_URL")
)

# IMPORTANT: Service Role Key is different from Publishable Key!
# Service Role Key is needed for admin operations like creating users
SUPABASE_SERVICE_ROLE_KEY = (
    os.getenv("SUPABASE_SERVICE_ROLE_KEY") or
    os.getenv("SUPABASE_SERVICE_KEY")
)

if not SUPABASE_URL:
    print("❌ Error: SUPABASE_URL not found")
    print("\nSet it in your .env file or environment:")
    print("  VITE_SUPABASE_URL='https://your-project.supabase.co'")
    print("\nOr export it:")
    print("  export VITE_SUPABASE_URL='https://your-project.supabase.co'")
    sys.exit(1)

if not SUPABASE_SERVICE_ROLE_KEY:
    print("❌ Error: SUPABASE_SERVICE_ROLE_KEY not found")
    print("\n⚠️  IMPORTANT: You need the SERVICE ROLE KEY (not publishable key)!")
    print("\nTo get your Service Role Key:")
    print("  1. Go to Supabase Dashboard")
    print("  2. Settings → API")
    print("  3. Find 'service_role' key (it's a SECRET key, starts with 'eyJ...')")
    print("  4. Copy it and add to your .env file:")
    print("     SUPABASE_SERVICE_ROLE_KEY='your-service-role-key-here'")
    print("\nOr export it:")
    print("  export SUPABASE_SERVICE_ROLE_KEY='your-service-role-key'")
    print("\n⚠️  WARNING: Never commit the service role key to git!")
    sys.exit(1)

def read_profiles_csv():
    """Read profiles from CSV file"""
    profiles = []
    try:
        with open('profiles.csv', 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                profiles.append({
                    'id': row['id'],
                    'email': row['email'],
                    'full_name': row['full_name'],
                    'role': row['role']
                })
    except FileNotFoundError:
        print("❌ Error: profiles.csv not found")
        print("   Run 'python generate_sample_data.py' first to generate it")
        sys.exit(1)
    return profiles

def create_user_via_api(profile, supabase_url, service_role_key):
    """Create user via direct HTTP API call (more reliable)"""
    import requests
    
    url = f"{supabase_url}/auth/v1/admin/users"
    
    headers = {
        "apikey": service_role_key,
        "Authorization": f"Bearer {service_role_key}",
        "Content-Type": "application/json",
    }
    
    payload = {
        "email": profile['email'],
        "password": "SamplePassword123!",
        "email_confirm": True,
        "user_metadata": {
            "full_name": profile['full_name'],
            "role": profile['role']
        }
    }
    
    # Try with ID first
    try:
        payload_with_id = {**payload, "id": profile['id']}
        response = requests.post(url, headers=headers, json=payload_with_id, timeout=10)
        if response.status_code in [200, 201]:
            return response
        # If 422, try without ID
        if response.status_code == 422:
            response = requests.post(url, headers=headers, json=payload, timeout=10)
            return response
        return response
    except requests.exceptions.RequestException as e:
        class MockResponse:
            def __init__(self):
                self.status_code = 500
                self.text = f"Network error: {str(e)}"
        return MockResponse()

def create_auth_users():
    """Create auth users in Supabase"""
    profiles = read_profiles_csv()
    print(f"📋 Found {len(profiles)} profiles in profiles.csv")
    print("🚀 Creating auth users...\n")
    print(f"🔗 Using Supabase URL: {SUPABASE_URL[:30]}...")
    print(f"🔑 Service Role Key: {SUPABASE_SERVICE_ROLE_KEY[:20]}...\n")
    
    created = 0
    skipped = 0
    errors = 0
    
    # Always use direct API calls (more reliable)
    import requests
    print("📡 Using direct HTTP API calls to Supabase Admin API...\n")
    
    # Test the connection first
    print("🔍 Testing service role key...")
    test_url = f"{SUPABASE_URL}/auth/v1/admin/users?limit=1"
    test_headers = {
        "apikey": SUPABASE_SERVICE_ROLE_KEY,
        "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE_KEY}",
    }
    try:
        test_response = requests.get(test_url, headers=test_headers, timeout=5)
        print(f"   Test response: {test_response.status_code}")
        if test_response.status_code == 200:
            print("✅ Service role key is valid and has admin access\n")
        elif test_response.status_code == 401 or test_response.status_code == 403:
            print("❌ ERROR: Service role key is invalid or doesn't have admin permissions")
            print("\n🔍 Debugging info:")
            print(f"   URL: {test_url}")
            print(f"   Key length: {len(SUPABASE_SERVICE_ROLE_KEY)} characters")
            print(f"   Key starts with: {SUPABASE_SERVICE_ROLE_KEY[:30]}...")
            print("\n💡 Please verify:")
            print("   1. You're using SERVICE_ROLE_SECRET (not anon or publishable key)")
            print("   2. From: Supabase Dashboard → Settings → API → service_role (secret)")
            print("   3. The key is the FULL secret (usually 200+ characters)")
            print("\n⚠️  If this persists, use the SQL alternative (see create_auth_users.sql)")
            sys.exit(1)
        else:
            print(f"⚠️  Warning: Test request returned {test_response.status_code}")
            if test_response.text:
                print(f"   Response: {test_response.text[:200]}")
            print("   Continuing anyway...\n")
    except Exception as e:
        print(f"⚠️  Could not test connection: {e}")
        print("   Continuing anyway...\n")
    
    for i, profile in enumerate(profiles, 1):
        try:
            response = create_user_via_api(profile, SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)
            
            if response.status_code == 200 or response.status_code == 201:
                created += 1
                print(f"✅ [{i}/{len(profiles)}] Created user: {profile['email']}")
            elif response.status_code == 422:
                # User already exists or validation error
                try:
                    error_data = response.json() if response.text else {}
                    error_str = json.dumps(error_data) if isinstance(error_data, dict) else str(error_data)
                    if "already" in error_str.lower() or "duplicate" in error_str.lower():
                        skipped += 1
                        print(f"⚠️  [{i}/{len(profiles)}] User already exists: {profile['email']}")
                    else:
                        errors += 1
                        print(f"❌ [{i}/{len(profiles)}] Validation error: {error_str[:150]}")
                except:
                    errors += 1
                    print(f"❌ [{i}/{len(profiles)}] Validation error (422): {response.text[:150]}")
            elif response.status_code == 401 or response.status_code == 403:
                errors += 1
                print(f"❌ [{i}/{len(profiles)}] Permission denied ({response.status_code})")
                print(f"   ⚠️  Check your SERVICE_ROLE_SECRET key is correct")
                if i == 1:  # Only show this once
                    print(f"   URL: {SUPABASE_URL}/auth/v1/admin/users")
                    print(f"   Key starts with: {SUPABASE_SERVICE_ROLE_KEY[:20]}...")
            else:
                errors += 1
                error_msg = response.text[:200] if hasattr(response, 'text') and response.text else f"HTTP {response.status_code}"
                print(f"❌ [{i}/{len(profiles)}] Error ({response.status_code}): {error_msg}")
                
        except Exception as e:
            errors += 1
            print(f"❌ [{i}/{len(profiles)}] Exception: {str(e)}")
    
    print(f"\n📊 Summary:")
    print(f"   ✅ Created: {created}")
    print(f"   ⚠️  Skipped (already exists): {skipped}")
    print(f"   ❌ Errors: {errors}")
    
    if created > 0:
        print(f"\n✅ Successfully created {created} auth users!")
        print("💡 Now you can import profiles.csv to Supabase")
        print("💡 Default password for all users: SamplePassword123!")
    elif skipped > 0 and errors == 0:
        print(f"\n⚠️  All users already exist. You can proceed to import profiles.csv")
    elif errors > 0:
        print(f"\n❌ Failed to create users. {errors} errors occurred.")
        print("\n🔍 Troubleshooting:")
        print("   1. Verify you're using SERVICE_ROLE_KEY (not publishable key)")
        print("   2. Check Supabase Dashboard → Settings → API → service_role key")
        print("   3. Ensure the key starts with 'eyJ' and is the full secret key")
        print("   4. Check if your Supabase project has user creation restrictions")
        print("   5. Try creating one user manually in Supabase Dashboard to test")
        print("\n💡 Alternative: Use SQL script (create_auth_users.sql) with admin access")

if __name__ == "__main__":
    create_auth_users()

File: create_auth_users.js
// Run this script with Node.js to create users properly
// npm install @supabase/supabase-js csv-parser
// Requires Node.js 18+ (for native fetch support)

import { createClient } from '@supabase/supabase-js';
import fs from 'fs';
import csv from 'csv-parser';

// Node.js 18+ has native fetch, so we can use it directly
if (typeof fetch === 'undefined') {
  console.error('❌ Error: fetch is not available. Please use Node.js 18+');
  console.error('   Or install node-fetch: npm install node-fetch');
  process.exit(1);
}

// Replace with your actual values
const SUPABASE_URL = process.env.SUPABASE_URL || 'https://kamllplbsvxwgguqitfe.supabase.co';
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImthbWxscGxic3Z4d2dndXFpdGZlIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc2MjcxMTc1NSwiZXhwIjoyMDc4Mjg3NzU1fQ.OK6DaZO61DnX5JG_68y5QTC6T10b4iHmTgkGT5al710';

// Validate that Supabase credentials are set
if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY || 
    SUPABASE_URL.includes('your-project') || 
    SUPABASE_SERVICE_ROLE_KEY.includes('your-service-role-key')) {
  console.error('❌ Error: Please set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY');
  console.error('   Either set environment variables or edit this file');
  console.error('\n   Example:');
  console.error('   export SUPABASE_URL="https://your-project.supabase.co"');
  console.error('   export SUPABASE_SERVICE_ROLE_KEY="your-service-role-key"');
  console.error('   node create_auth_users.js');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
});

const PASSWORD = 'SamplePassword123!';

async function readProfilesFromCSV() {
  return new Promise((resolve, reject) => {
    const profiles = [];
    fs.createReadStream('profiles.csv')
      .pipe(csv())
      .on('data', (row) => {
        profiles.push({
          id: row.id,
          email: row.email,
          full_name: row.full_name,
          role: row.role
        });
      })
      .on('end', () => {
        resolve(profiles);
      })
      .on('error', (error) => {
        reject(error);
      });
  });
}

async function createUser(profile, retryCount = 0) {
  const maxRetries = 3;
  const retryDelay = 2000; // 2 seconds
  
  try {
    // Use direct REST API call to create user with specific ID
    // This requires the service role key and allows setting custom IDs
    const response = await fetch(`${SUPABASE_URL}/auth/v1/admin/users`, {
      method: 'POST',
      headers: {
        'apikey': SUPABASE_SERVICE_ROLE_KEY,
        'Authorization': `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        id: profile.id,
        email: profile.email,
        password: PASSWORD,
        email_confirm: true,
        user_metadata: {
          full_name: profile.full_name,
          role: profile.role
        }
      })
    });

    const result = await response.json();

    if (!response.ok) {
      // Check if user already exists
      if (response.status === 422 || (result.error && (result.error.includes('already') || result.error.includes('duplicate')))) {
        console.log(`⚠️  User already exists: ${profile.email}`);
        return { success: true, skipped: true, user: null };
      }
      
      // Retry on 500 errors (server errors)
      if (response.status === 500 && retryCount < maxRetries) {
        console.log(`⚠️  Server error (HTTP 500) for ${profile.email}, retrying in ${retryDelay/1000}s... (${retryCount + 1}/${maxRetries})`);
        await new Promise(resolve => setTimeout(resolve, retryDelay));
        return createUser(profile, retryCount + 1);
      }
      
      // Show detailed error
      const errorMsg = result.error || result.message || `HTTP ${response.status}`;
      console.error(`❌ Error creating ${profile.email}: ${errorMsg}`);
      if (result.error_description) {
        console.error(`   Details: ${result.error_description}`);
      }
      return { success: false, error: new Error(errorMsg) };
    }

    if (result.id === profile.id) {
      console.log(`✅ Created ${profile.email} with ID: ${result.id}`);
      return { success: true, user: result };
    } else {
      console.log(`✅ Created ${profile.email} with ID: ${result.id} (expected: ${profile.id})`);
      return { success: true, user: result, idMismatch: true };
    }
  } catch (err) {
    // Retry on network errors
    if (retryCount < maxRetries && (err.message.includes('fetch') || err.message.includes('network'))) {
      console.log(`⚠️  Network error for ${profile.email}, retrying in ${retryDelay/1000}s... (${retryCount + 1}/${maxRetries})`);
      await new Promise(resolve => setTimeout(resolve, retryDelay));
      return createUser(profile, retryCount + 1);
    }
    
    console.error(`❌ Error creating ${profile.email}:`, err.message);
    return { success: false, error: err };
  }
}

async function createUsers() {
  console.log('📋 Reading profiles from profiles.csv...\n');
  
  let profiles;
  try {
    profiles = await readProfilesFromCSV();
    console.log(`Found ${profiles.length} profiles to create\n`);
  } catch (error) {
    console.error('❌ Error reading profiles.csv:', error.message);
    console.error('   Make sure profiles.csv exists in the current directory');
    process.exit(1);
  }

  console.log('🚀 Creating auth users...\n');
  
  let created = 0;
  let skipped = 0;
  let errors = 0;
  let idMismatches = 0;

  for (const profile of profiles) {
    const result = await createUser(profile);
    
    if (result.success) {
      if (result.skipped) {
        skipped++;
      } else {
        created++;
        if (result.idMismatch) {
          idMismatches++;
        }
      }
    } else {
      errors++;
    }
    
    // Longer delay to avoid rate limiting (HTTP 500 errors)
    await new Promise(resolve => setTimeout(resolve, 500));
  }

  console.log('\n📊 Summary:');
  console.log(`   ✅ Created: ${created}`);
  console.log(`   ⚠️  Skipped (already exists): ${skipped}`);
  console.log(`   ❌ Errors: ${errors}`);
  if (idMismatches > 0) {
    console.log(`   ⚠️  ID Mismatches: ${idMismatches} (see warnings above)`);
  }

  if (created > 0 || skipped > 0) {
    console.log('\n✅ Users are ready! You can now login with:');
    console.log('   Email: user2@example.com');
    console.log('   Password: SamplePassword123!');
  }
}

// Run the script
createUsers().catch((error) => {
  console.error('\n❌ Fatal error:', error);
  process.exit(1);
});

